R/roc.R

```diff
@@ -0,0 +1,261 @@
+                         cutoffs_from = NULL,
+                         cutoffs_to = NULL,
+                         mod_names = c("par", "str"), 
+                         alpha = NULL,
+                         lambda = NULL, 
+                         theta = NULL,
+                         psi = NULL,
+                         nu = NULL, ...) {
+  
+  if(!missing(cfa_fit)) {
+    est <- format_cfa_partinv(cfa_fit, comp = "est")
+    n_g <- cfa_fit@Data@ngroups # number of groups
+    psi <- est$psi
+    lambda <- est$lambda
+    theta <- est$theta
+    alpha <- est$alpha
+    nu <- est$nu
+  }
+  if(missing(cfa_fit) & !is.null(alpha) & !is.null(psi) & 
+     !is.null(lambda) & !is.null(theta)  & !is.null(nu)) {
+    n_g <- length(alpha) # number of groups
+  }
+  
+  propsels <- seq(from = from, to = to, by = by)
+  use <- "propsels"
+  xl <- "Proportion of selection"
+  rangeVals <- propsels
+  
+  if ((is.null(cutoffs_from) && !is.null(cutoffs_to)) ||
+      (!is.null(cutoffs_from) && is.null(cutoffs_to))) {
+  }
+  
+  # if the user provided the max and min cutoff values, update rangeVals with 
+  # a range of cutoffs
+  if (!is.null(cutoffs_from) && !is.null(cutoffs_to)) {
+    cutoffs <- seq(from = cutoffs_from, to = cutoffs_to, by = by)
+    rangeVals <- cutoffs
+    xl <- "Thresholds" # for the plots later
+    use <- "cutoffs"
+  }
+  
+  
+  # if the user did not provide labels, or provided the wrong number of labels,
+  if ((is.null(labels) || (length(labels) != n_g)) && !missing(cfa_fit)) {
+    labels <- cfa_fit@Data@group.label
+    labels <- paste(labels, c("(reference)", rep("(focal)", n_g - 1)))
+  }
+  if ((is.null(labels) || (length(labels) != n_g)) && missing(cfa_fit)) {
+    labels <- paste(rep("Group"), seq(1:n_g))
+    labels <- paste(labels, c("(reference)", rep("(focal)", n_g - 1)))
+  }
+  
+  cai <- c("Sensitivity", "Specificity")
+  ls_mat <- matrix(NA, ncol = length(rangeVals), nrow = n_g,
+                   dimnames = list(labels, rangeVals))
+  ls_names <- c(t(outer(cai, Y = mod_names, FUN = paste, sep = "_")))
+  vals <- rep(list(ls_mat), length(ls_names))
+  names(vals) <- ls_names
+  
+  # if pmix is missing, assume equal mixing proportions
+  if (is.null(pmix)) pmix <- as.matrix(c(rep(1 / n_g, n_g)), ncol = n_g)
+  pmix <- as.vector(pmix)
+  
+  ylabs <- ""
+  mains <- ""
+  
+  # call PartInv with each proportion of selection and store CAI in the list of
+  # data frames
+  for (p in seq_along(rangeVals)) {
+    # if the user provided cutoff values
+    if (use == "cutoffs") {
+      suppressWarnings({
+        pinv <- PartInv(cut_z = cutoffs[p],
+                        psi = psi,
+                        lambda = lambda,
+                        theta = theta,
+                        alpha = alpha,
+                        nu = nu,
+                        pmix = pmix,
+                        plot_contour = FALSE,
+                        labels = labels,
+                        show_mi_result = TRUE)
```

Comment:
Is it possible to call `compute_cai()` instead of `PartInv()` here? Not sure if the former can already to the job of getting SE and SP.

Reply:
@marklhc I've spent some time working on updating the code today to use `compute_cai()` instead, but such a change requires a number of changes in different helper functions and some additional updates, and there is a mismatch between how the functions are structured. I'd like to keep it as `PartInv` for now if that's alright with you, since it seems like a more extensive update than it initially appeared. Please let me know what you think!